### PR TITLE
get all composite key candidates function added

### PR DESCRIPTION
**Pull request by souvik-databricks**
Added the function to get all of the possible composite key candidates for a delta table

### CODE REVIEW

File: mack/__init__.py [hunk 1]
```diff
@@ -384,6 +384,33 @@ def find_composite_key_candidates(
...
 
+def find_all_composite_key_combos(
+    df: Union[DeltaTable, DataFrame], exclude_cols: List[str] = None
+):
+    if type(df) == DeltaTable:
+        df = df.toDF()
+    if exclude_cols is None:
+        exclude_cols = []
+    df_col_excluded = df.drop(*exclude_cols)
+    col_select_condition = df_col_excluded.distinct().count()
```
Comment (robertkossendey) [inline mack/__init__.py:395]: @souvik-databricks why does it need to distinct? If .distinct().count() < .count(), there is no unique combination, or?

Comment (souvik-databricks) [reply mack/__init__.py:395]: @robertkossendey .distinct() is required because if your data has duplicates in them and then the collect_set() in line 401 will matchup and we have  only unique combination of values for the columns.

Comment (robertkossendey) [reply mack/__init__.py:395]: @souvik-databricks still struggling to understand, could you add a test with a completely duplicate line? Because that should return no combination at all right?

Comment (souvik-databricks) [reply mack/__init__.py:395]: @robertkossendey So in that scenario if you have a case of duplicate record then that would indicate that from the columns that you have included to search in, you don't have valid composite key combos. You would have to include more columns to act as part of the composite key.

If in the whole dataset you have a full exact duplicate record then it won't make any sense to keep it as that's then redundant information. 

And I will add a test with duplicate record with the explanation of the output for that test case in the readme.

Comment (robertkossendey) [reply mack/__init__.py:395]: > If in the whole dataset you have a full exact duplicate record then it won't make any sense to keep it as that's then redundant information.

@souvik-databricks but isn't that the information that there is no valid key candidate in the provided dataset?

**Comment by MrPowers** [comment]
This looks like a great start.  Can you add more tests for the edge cases like when the result is empty, when the DataFrame contains null values, etc?

File: mack/__init__.py [hunk 2]
```diff
@@ -384,6 +384,33 @@ def find_composite_key_candidates(
...
+    exprs = [size(collect_set(x)).alias(x) for x in finalcols]
+    df_col_excluded = df_col_excluded \
+        .withColumn("column_combos ->", lit("distinct_row_counts ->")) \
+        .groupBy("column_combos ->") \
+        .agg(*exprs)
+    columns = [
+        column for column in df_col_excluded.columns if
+        df_col_excluded.select(column).collect()[0][0] == col_select_condition
+    ]
+    df_col_excluded.select("column_combos ->", *columns).show(truncate=False)
```
Comment (robertkossendey) [inline mack/__init__.py:410]: @souvik-databricks we can remove that show, right?

File: mack/__init__.py (hunk 1 above)
Comment (souvik-databricks) [reply mack/__init__.py:395]: I was OOO for last week. @robertkossendey I am back now, I will resume the work now.

**Comment by MrPowers** [comment]
@souvik-databricks - let me know when you've addressed the feedback from @robertkossendey and this is ready for another review, thank you!

**Comment by MrPowers** [comment]
@souvik-databricks - following up on this one.  Are you planning to address this?

**Comment by souvik-databricks** [comment]
@MrPowers Let's go through this details over our sync.

**Comment by MrPowers** [comment]
@souvik-databricks - anything else we need to push this over the finish line?
